### ISSUE TITLE
CPU usage increases over time until the UI becomes unresponsive. [mac]

### ISSUE THREAD

**Issue by pianojoe** [jamulussoftware/jamulus#1643]
On Mac, after two hours of rehearsal with ~40 participants, the UI becomes unresponsive. A click may take 20s to execute. `top` shows a processor usage of 103%. (As compared to 35% one hour earlier, with the same number of participants.)
The Mac mouse cursor changes into the "spinning beachball of death", as it is commonly reffered to.

CPU usage shouldn’t grow over time. 

I’m on macos 11.3.1 on a 2.9 GHz 6-Core Intel Core i9 MacBookPro (definitely not underpowered). I’m using a USB audio interface Focusrite Scarlett 4i4.

The behaviour was observed on Jamulus 3.7.0 (obtained as a .dmg from jamulus.io) AND on Jamulus 3.8.0beta1 (which I compiled myself).

**Comment by hoffie** [comment]
Thanks for the report.

1. Just to confirm, you are talking about the Jamulus client, right?

2. Just a blind guess because it seems to be the reason for another recent Mac-related issue: Which version of Qt did you use on your local build? The official `.dmg` is built against rather old Qt (5.9) which seems to cause problems (UI glitches) on recent Mac versions.
Could you try this build and report back if this improves things for you?
https://github.com/jamulussoftware/jamulus/issues/1628#issuecomment-832859835

3. Does memory usage also increase?

**Comment by pianojoe** [comment]
hoffie, 

1. Yes, I’m talking about the Jamulus Client.
2. Qt 5.15.2. This is the very same Qt version as your build.
3. Need to check. I’ll get back to you on this.

**Comment by hoffie** [comment]
Ok. My only idea for getting close to the cause of the issue would be to try different working/non-working versions. Does 3.6.2 exhibit the same problem? If not, that would be something to start from. Are you familiar with git bisect? That way you could generate different builds and check if they are broken or not. This will likely require some iterations of building / verifying / building again, etc., but will lead to a single commit which is the cause for the change in behavior.

This requires effort, but I don't know where else to start. :(

**Comment by softins** [comment]
Also, when it gets into this state, does clicking Disconnect and reconnecting give any improvement, or is it necessary to exit and restart the program?

**Comment by MattMerz-Violin** [comment]
I have the same problem. Jamulus 3.7 becomes unresponsive after playing for some time and not touching any UI item. It then takes up to 20 seconds until the mouse click is recognized and the UI is reacting again. MacOS 11.3.1.

I use the official Jamulus for Mac version that can be downloaded from SourceForge.

**Comment by softins** [comment]
For the first time, I experienced this tonight. Macbook Pro on Mojave, running the 3.8.0beta1 release.

This was after playing with four others for about 90 minutes.

Interestingly, the audio continued without significant delay, although the ping and delay stopped updating, and attempting to use the GUI gave the spinning beach ball. This prevented disconnecting.

Running `top` in a terminal showed Jamulus using over 90% of the CPU.

I quit and restarted Jamulus. I have a slight suspicion that using the menu bar might have unstuck it, but I had already clicked on Quit.

Will investigate more...

**Comment by hoffie** [comment]
Good to know that this is reproducible with just five participants which should be easier to simulate than 40.
I've ran multiple 90+ minute sessions with more than 4 people with both 3.7.0 and ~3.8.0beta1 on Linux, so this is either really Mac-specific (as suspected) or needs another factor to become an issue.

**Comment by MattMerz-Violin** [comment]
Hi, a few answers (more to come):

What is the minimal number of users to trigger this for you? MM: happens for me with 3 Players and also with up to 20 (did not try with more)
What is the minimal number of minutes after this happens? MM: will try to find out
Can you answer @softins's question regarding the Disconnect test above? MM: as soon as the program reacts again, it reacts totally normal again without restarting. 
Can you try with an empty .ini file (e.g. start via Jamulus --inifile /tmp/empty.ini or locate the default ini and rename/remove it)? MM: will do so and Report.
If you are familiar with building and using git, it would be really helpful to run the steps from #1643 (comment) MM: I am not familiar with building programs myself :-(

**Comment by softins** [comment]
Please could those affected by this issue try the macOS build from the new [3.8.0beta2 version](https://github.com/jamulussoftware/jamulus/releases/tag/r3_8_0beta2)?

It has come to light that both 3.7.0 and 3.8.0beta1 were built using Apple SDK 11.1, which Qt5 claims is not fully compatible. So it's possible, although not certain, that could be related to this issue.

See PR #1655 for more information.

**Comment by MattMerz-Violin** [comment]
Hi, I tried 3.8 beta 2 today in a 2 hour Jamulus session and it still shows the same behaviour under MacOS 11.3.1. After a certain time 30 to 60 minutes in the background, the app becomes unresponsive when I click again in the Jamulus window. Then the spinning wheel shows for 10 or 15 seconds an then Jamulus reacts again normal to every input.

**Comment by JW144754** [comment]
now that this is brought up, I have totally have had this experience before. During the school year we had2 hours rehearsals with 40+ people twice a week, and i do not know about everyone in the ensemble, but I know that a couple of my friends who are on Mac would need to disconnect and reconnect during the break at 1 hour, otherwise the clicks becomes VERY irresponsible, ie clicking on "Solo" might take 10 seconds to actually react and solo the track, and same with un-soloing.

I just assumed that this is just because of large ensemble and extended time, so that is why I never bothered reporting it. I will ask the ensemble and come back here.

**Comment by Sommi486** [comment]
The behavior ist still in 3.8.0rc1

After 8 minutes not touching the client an sending no sound the client ist frozen.

**Comment by softins** [comment]
> The behavior ist still in 3.8.0rc1
> 
> After 8 minutes not touching the client an sending no sound the client ist frozen. Happened with Catalina and BigSur

Was this while connected to a server, or not?

**Comment by Sommi486** [comment]
Yes, the client was connected with a server. 

Server also running 3.8.0rc1

**Comment by softins** [comment]
We would definitely like to get to the bottom of this issue, although it might take some time, and will likely be after the 3.8.0 release. It appears to have been observed in a few different versions.

This kind of issue, that takes an indeterminate time to appear, is difficult to track down, and to help us identify any pattern, any reports need to include as much detail as possible, such as:
- version of macOS and version of Jamulus
- how many clients were connected to the session
- how long (approx) the session had been going
- how long (approx) since the last interaction with the program (i.e. Does interacting with the GUI keep the problem at bay? Does it only happen after leaving the GUI idle for a time?)
- was the Settings window open (displaying ping time, etc)?
- If so, did the ping and delay times continue to update or freeze?
- Did the sound continue during the time the GUI was frozen?
- Did the GUI unfreeze itself after a while?
- Did some other operation unfreeze the GUI, e.g. selecting an option in the menu bar?
- Was it necessary to kill the program?
- If the chat window was open, did Chat work for sending or receiving?
- Anything else that could conceivably be relevant.

**Comment by softins** [comment]
It happened to me three times this evening, using 3.8.0rc1 on macOS Mojave. On each occasion, the audio was not interrupted, but I had not touched the GUI for a while.

On the first occasion, the first thing I noticed was erratic values being displayed for Ping and Delay, despite not experiencing any significant delay or disturbance to the audio. Thinking this was a different problem, I went to disconnect and reconnect, but the Disconnect button did not respond, and I got the spinning beachball. I tried interacting with the menus and the windows, but without success, and I needed to do Force Quit.

On the second occasion, I saw the beachball, and the ping and delay values were not updating. After trying various things, I moved the actual Jamulus window. That immediately solved the issue and normal operation was resumed.

Some time later I found the same thing, but in this case, moving the window did not immediately restore normal operation. But after fiddling for a short while moving windows etc., it did un-freeze again and resume normal operation.

**Comment by MattMerz-Violin** [comment]
Had a session yesterday with 18 singers. Jamulus did not respond after a certain time (30 to 45 minutes approximately). 

The network information was showing strange numbers but the session was running perfectly, the beachball was spinning for about 2 or 3 minutes, the GUI was not reacting at all, we were continuing to sing all the time without any problem. For me it sounds like its mainly a GUI problem, the part doing the networking and all the music part works perfectly.

![Jamulus frozen 3 8rc1 macos 11 3 1](https://user-images.githubusercontent.com/83976915/119083714-c3a3f280-ba00-11eb-9923-27ed8410abd3.jpg)

**Comment by softins** [comment]
OK, it seems what we are looking for is a memory leak. I have been monitoring the Jamulus process using `ps -cvp $pid`, and when the program is first started up and not connected to a server, the memory usage is stable:
```
2021-05-21 16:05:55
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:01.93   0   0      0  4959052 102524     -        0   0.0  1.2 Jamulus

2021-05-21 16:06:00
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:01.94   0   0      0  4959052 102524     -        0   0.0  1.2 Jamulus
```
The VSZ (virtual size) and RSS (resident set size) are in K, so this is about 4.9G of virtual image (probably most of that is the Qt libraries), and about 1G of real RAM usage.

Once it is connected to a server, the RAM usage starts climbing steadily:
```
2021-05-21 16:08:20
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:05.76   0   0      0  5119164 174364     -        0  10.0  2.1 Jamulus

2021-05-21 16:08:25
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:06.31   0   0      0  5118640 175636     -        0   9.9  2.1 Jamulus

2021-05-21 16:08:30
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:06.90   0   0      0  5118640 177116     -        0  13.1  2.1 Jamulus

2021-05-21 16:08:35
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:07.63   0   0      0  5116544 178860     -        0   8.8  2.1 Jamulus

2021-05-21 16:08:40
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:08.10   0   0      0  5116544 179720     -        0  19.9  2.1 Jamulus

2021-05-21 16:08:45
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 R      0:08.71   0   0      0  5116544 181408     -        0  25.0  2.2 Jamulus

2021-05-21 16:08:50
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 R      0:09.74   0   0      0  5117760 182160     -        0  19.8  2.2 Jamulus

2021-05-21 16:08:55
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:10.42   0   0      0  5116712 182184     -        0  10.9  2.2 Jamulus

2021-05-21 16:09:00
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      0:11.06   0   0      0  5116592 182096     -        0   6.5  2.2 Jamulus
```
Eventually, the free RAM gets exhausted (my Macbook has 8G of RAM) and the system starts swapping. That is probably when the network stats start jumping around. They did so even when connected to a server on my LAN.

While the program is being interacted with, the parts of the image that are needed for the GUI are probably kept in RAM and the leaked memory is swapped out. But if the GUI is left alone for an extended period, it seems that a LOT of needed GUI code is swapped out, and the spinning beachball is shown while the system goes about swapping it all back in again. During this time the audio continues running, because the audio code had been active all the time and was not swapped out.

This was the `ps` display around when the network stats first started jumping:
```
2021-05-21 16:22:20
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 R      2:41.45   0   0      0  5718752 425732     -        0  28.3  5.1 Jamulus

2021-05-21 16:22:25
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 S      2:42.66   0   0      0  5718752 427100     -        0  25.7  5.1 Jamulus

2021-05-21 16:22:30
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 R      2:47.32   0   0      0  5718752 427400     -        0 103.0  5.1 Jamulus

2021-05-21 16:22:35
  PID STAT      TIME  SL  RE PAGEIN      VSZ    RSS   LIM     TSIZ  %CPU %MEM COMMAND
43653 R      2:52.47   0   0      0  5716704 425352     -        0 103.1  5.1 Jamulus
```

Between 16:22:25 and 16:22:30 was when I clicked on the GUI and git the spinning beachball. The 100% CPU is while it tries to swap everything back. In this instance it only took a few seconds. If I had left it a lot longer, it would have taken more time.

I'll perform the same tests on a Linux client. If it also shows memory leakage, that will be an easier platform on which to debug.

So now on the trail of memory leaks. Maybe a needle in a haystack, but hopefully not!

**Comment by softins** [comment]
Well monitoring the 3.8.0rc1 client running on a Pi does not show a steadily increasing RSS while connected.

So that suggests the memory leak may be Mac specific. First place to go and look is the Mac sound handling code.

**Comment by softins** [comment]
Notwithstanding the [comment above](https://github.com/jamulussoftware/jamulus/issues/1643#issue-877696323), I've just tested 3.6.2, 3.7.0 and 3.8.0beta1 on the Mac. The quickly rising memory usage does not occur with 3.6.2 or 3.7.0, but does happen with 3.8.0beta1. I guess it's possible there is a much slower leak on 3.7.0, but let's find the fast one first!

**Comment by MattMerz-Violin** [comment]
Is there an easy possibility to support the search for the leak? I will play on Jamulus tomorrow some sessions and could help, in case there is an easy way to do that like running something in parallel etc. I am not a programer at all, only a user.

**Comment by softins** [comment]
> Is there an easy possibility to support the search for the leak? I will play on Jamulus tomorrow some sessions and could help, in case there is an easy way to do that like running something in parallel etc. I am not a programer at all, only a user. 

Thanks for the offer, Matt, but I don't think there is anything you could do at present. I have to gather some information from different versions and build environments, to try and narrow down in which version the problem started. Then it will be a case of studying the changes introduced in that version.

**Comment by softins** [comment]
Hmm, looking again at the figures, I'm not sure it is as simple as a memory leak. The RSS is an order of magnitued smaller than I thought at first glance. Also, after leaving it running with the RSS gradually increasing, and then interacting with the GUI, once the CPU stops being at 100% and the beachball disappears, the RSS does in fact drop back quite a way. So maybe there is some kind of garbage collection happening within Qt? Still puzzling...

**Comment by hoffie** [comment]
Just some other ideas:
- Is there a difference between self-built 3.6.2 (or 3.7.0) vs. the official releases by corrados (3.6.2) / Github (3.7.0)?
- Is this reproducible on a non-GUI client (`Jamulus -c localhost -n`)? That should rule out all the GUI stuff.
- Is it reproducible on a GUI server? That rule out any audio input/output handling stuff.

As said in the chat, `mac/` doesn't seem to have seen any code changes (despite coding style changes) between 3.7.0 and 3.8.0rc1.

Regarding severity: If we are sure that 3.8.0rc1 is worse than 3.7.0 we may consider delaying the upcoming release. If this is not the case, we should move forward with the release and publish a bugfix release as soon as the bug is tracked down and fixed.

**Comment by MattMerz-Violin** [comment]
Hi, I tried Jamulus 3.5.12 under MacOS 11.3.1 and monitored the RAM usage and there is no increase in RAM usage if Jamulus runs in the background, connected to a server and playing music. It uses between 80 an 100MB RAM.

When I did the same setting with 3.8 RC1m(main window and settings window open in Jamulus, connected to the same server, playing the same music, same audio interface setting etc.) Ram usage increases as soon as Jamulus runs in the background. I stopped it, when it reached over 400MB of RAM used (started at somewhere between 90 and 120 MB after I started Jamulus).
When switching back to the Jamulus Window, the RAM usage dropped to the initial values.

So it looks like the bug came in after version 3.5.12. I hope this is useful information.

**Comment by dcorson-ticino-com** [comment]
So you are saying this only happens when Jamulus is not the foreground application ?

**Comment by softins** [comment]
I have been doing git bisect testing today with the client always running foreground, and later versions exhibit the problem while earlier versions don’t. Still homing in...

**Comment by MattMerz-Violin** [comment]
I played today 2.5h in a session with Jamulus 3.7 on MacOS 11.3.1 and I always kept Jamulus in the foreground. I had no issues at all when doing that. I switched to the browser some times to look for sheet music, but I always went back directly to put Jamulus into foreground and no issues with non-reacting GUI.

**Comment by softins** [comment]
I have today done carefully controlled tests using `git bisect`, testing each build under identical conditions for 30 minutes with a script measuring and logging the memory usage every 5 seconds and computing statistics on the VSZ and RSS usage, and their changes over a sliding 5-minute window.

The evidence is that a memory leak was introduced by [this commit](https://github.com/jamulussoftware/jamulus/commit/d94fd69e). We have yet to examine the code to identify the reason, or to understand why it only seems to affect Mac. That is the next step.

This is the output from the test script for the above commit:
```
Commit = d94fd69e
Elapsed = 1800 sec, Periods = 360
VSZ: start=4963156  end=5055552  change=+92396  mean=4995834  median=4992880 | 5min change min=-14364 mean=+16961 median=+16384 max=+42408
RSS: start=236916  end=331536  change=+94620  mean=272258  median=270248 | 5min change min=-10664 mean=+16860 median=+18708 max=+38596
```

This shows the memory usage increasing over the period.

In contrast, this is the output from the parent of the commit:
```
Commit = 4f6033a8
Elapsed = 1800 sec, Periods = 360
VSZ: start=4903232  end=4888364  change=-14868  mean=4885227  median=4887340 | 5min change min=-21140 mean=-340 median=+0 max=+15820
RSS: start=169424  end=154056  change=-15368  mean=151649  median=154048 | 5min change min=-20628 mean=-444 median=+0 max=+15304
```

This shows the memory usage being stable, on average, over the period.

All tests were done by compiling against Qt 5.9.9 on macOS Mojave, and running the built Jamulus with the settings window open to display the ping and delay stats, and the main Jamulus window on top above it.

**Comment by softins** [comment]
As an additional data point, the memory leak doesn't occur when running with the settings window closed, nor when running with the settings window open on a different tab. It only occurs when the tab is open that displays the ping and delay repeatedly.

**Comment by softins** [comment]
Well progress today has identified that calling `QLabel::setText()` repeatedly on a label with different text values causes the memory “leak”, at least on Mac in the new layout of the settings page. It might not be a complete leak, but may instead be lazy garbage collection. Just started to look at the source for `QLabel::setText()` in Qt5, which shows a potentially promising line of enquiry to investigate tomorrow. The beauty of open source!

**Comment by softins** [comment]
So the promising line of enquiry didn't come to anything. This is proving a hard problem to understand and solve.

But some further data from testing on the Mac:

- If the settings tab with the ping stats is shown, but the settings window is _not_ the top window (this is the usual case for most people), and the user does _not_ interact with the GUI for a time, the memory usage increases with the repeated updates of `lblPingTimeValue` and `lblOverallDelayValue`. When the user next tries to interact with the GUI after a long idle period, the system goes busy for a while before recovering. This is the original problem reported.
- However, if the settings tab _is_ left as the top window, even though not interacting with the GUI, the memory usage does not increase.

It would appear that when updating the label, there is an old data item (widget?) that is flagged for later reclaiming `deleteLater()` in an event loop, but that this event loop is not running when the window with the label being updated is not the top window with focus. It the runs when the user next tries to interact, but the discarded objects from the many queued updates (2 per second) take a significant time to be handled.

When we need is for them to continue being handled in the background even when the window is not on top.

Will probably need to ask on the Qt Forums.

**Comment by pljones** [comment]
There's a way to trigger the event loop... or am I thinking of C# or Java Swing...

**Comment by softins** [comment]
Also Stack Overflow: https://stackoverflow.com/questions/67676000/qlabelsettext-on-inactive-window-of-mac

**Comment by softins** [comment]
> There's a way to trigger the event loop... or am I thinking of C# or Java Swing...

Don't know...

> (And why is it Mac-specific..?)

For that matter, why did it only appear when switching to a tabbed dialog?

**Comment by pljones** [comment]
> For that matter, why did it only appear when switching to a tabbed dialog?

That adds an extra container layer around the widget, which may be causing the the issue (i.e. not calling the clean up) regularly...

**Comment by hoffie** [comment]
> There's a way to trigger the event loop... or am I thinking of C# or Java Swing... (And why is it Mac-specific..?)

[QCoreApplication::processEvents()](https://doc.qt.io/qt-5/qcoreapplication.html#processEvents)? Sounds like it could help.

If it does, we should probably guard this workaround to only activate on Mac (and possibly only when the window is in the background -- not sure if this easily detectable).

Could it also be an issue with (missing?) object parent associations?

**Comment by softins** [comment]
It's definitely something to do with `QTabWidget`. For a test, I manually edited the XML to change the `QTabWidget` to a horizontal layout, and each contained `QWidget` to an item, so that all the information was on a single settings dialog. I also removed the event handlers for the tab changes. The commit is 979249d. Compared with its parent, this build on Mac does not build up memory with deferred deletions, when run in the same way (settings window open behind main window).

I have a couple of other suggestions to try while still using `QTabWidget`, including the `QCoreApplication::processEvents()` mentioned above.

**Comment by dcorson-ticino-com** [comment]
So Qt is forgetting to do something in their QtTabWidget.  
But still, why only on Mac?  
I hope you can find a solution, I am sorry I can't do anything as I have no Mac hardware.

**Comment by pianojoe** [comment]
I just had a rehearsal with 32 participants. After 60 minutes, the UI in the main window became fairly unresponsive. „Disconnect“ took ~30 seconds to go through and gave me the SBOD („spinning beachball of death“).

The thing is: The settings window wasn’t open at all, so there must be another memory leak.

**Comment by softins** [comment]
> I just had a rehearsal with 32 participants. After 60 minutes, the UI in the main window became fairly unresponsive. „Disconnect“ took ~30 seconds to go through and gave me the SBOD („spinning beachball of death“).
> 
> The thing is: The settings window wasn’t open at all, so there must be another memory leak.

Hmm, ok. It's important for us to know which version you were using on this particular occasion. Did you have any other Jamulus window open, e.g. chat or connect? Was the main window with faders on top, or another app open over it?

**Comment by ann0see** [comment]
Pinned to gain more visibility to this issue. I think we should discuss alternatives (as @hoffie suggested). Although we can’t revert the commit in question easily, we might move the ping/delay status somewhere else. What about the Measurements box around the stats? Is this related?

**Comment by MattMerz-Violin** [comment]
Hi, just played one hour with RC1 under MacOS 11.4 and had the settings window closed. When trying to disconnect at the end of the session (3 players) the Beachball showed up and it took 10 to 15 seconds until the program reacted. The RAM usage had not increased. 

So for me it looks like the settings window has a big problem that can be reproduced easily, but there seems to be another issue with the program also causing the GUI not reacting.

**Comment by softins** [comment]
@MattMerz-Violin thanks for the report! I suspect that this is a long-standing issue deep within the Qt implementation on Mac. We will not be able to solve it (or even understand it) before the 3.8.0 release.
